astropy/wcs/utils.py

```diff
@@ -149,41 +150,109 @@ def wcs_to_celestial_frame(wcs):
+    onto the "plane of intermediate world coordinates" as defined in
+    `Greisen & Calabretta 2002, A&A, 395, 1061 <http://adsabs.harvard.edu/abs/2002A%26A...395.1061G>`_.
+
+    .. note::
+        This function is concerned **only** about the transformation
+        "image plane"->"projection plane" and **not** about the
+        transformation "celestial sphere"->"projection plane"->"image plane".
+        Therefore, this function ignores distortions arising due to
+        non-linear nature of most projections.
+
+    .. note::
+        In order to compute the area of pixels corresponding to celestial
+        axes only, this function uses the `~astropy.wcs.WCS.celestial` WCS
+        object of the input ``wcs``.  This is different from the
+        `~astropy.wcs.utils.proj_plane_pixel_scales` function
+        that computes the scales for the axes of the input WCS itself.
+
+    Parameters
+    ----------
+    wcs : `~astropy.wcs.WCS`
+        A world coordinate system object.
+
+    Returns
+    -------
+    area : float
+        Area (in the projection plane) of the pixel at ``CRPIX`` location.
+        The units of the returned result are the same as the units of
+        the `~astropy.wcs.Wcsprm.cdelt`, `~astropy.wcs.Wcsprm.crval`,
+        and `~astropy.wcs.Wcsprm.cd` for the celestial WCS and can be
+        obtained by inquiring the value of `~astropy.wcs.Wcsprm.cunit`
+        property of the `~astropy.wcs.WCS.celestial` WCS object.
 
     Raises
     ------
-    ValueError : if the pixels are nonsquare and ``allow_nonsquare==False``
+    ValueError
+        Pixel area is defined only for 2D pixels. Most likely the
+        `~astropy.wcs.Wcsprm.cd` matrix of the `~astropy.wcs.WCS.celestial`
+        WCS is not a square matrix of second order.
+
+    Notes
+    -----
+
+    Depending on the aplication, square root of the pixel area can be used to
+    represent a single pixel scale of an equivalent square pixel
+    whose area is equal to the area of a generally non-square pixel.
+
+    See Also
+    --------
+    astropy.wcs.utils.proj_plane_pixel_scales
+
     """
-    cwcs = inwcs.celestial
-    if cwcs.wcs.ctype[0][-3:] != 'CAR':
-        warnings.warn("Pixel sizes may very over the image for "
-                      "projection class {0}".format(cwcs.wcs.ctype[0][-3:]),
-                      AstropyUserWarning)
-    scale = (cwcs.pixel_scale_matrix**2).sum(axis=0)**0.5
-    if not np.allclose(scale[0],scale[1]):
-        if allow_nonsquare:
-            warnings.warn("Pixels are not square, using an average pixel scale")
-            return np.mean(scale)*u.deg
-        else:
-            raise ValueError("Pixels are not square: 'pixel scale' is ambiguous")
-    # Return a quantity: WCS always stores in degrees
-    return scale[0]*u.deg
+    psm = wcs.celestial.pixel_scale_matrix
+    if psm.shape != (2, 2):
+        raise ValueError("Pixel area is defined only for 2D pixels.")
+    return float(np.abs(np.linalg.det(psm)))
```

Comment:
is the `float` necessary here?  The pixel scales are already guaranteed to be floats, no?